### DISCUSSION TITLE
Help undestanding how to write an insert mutation with many-to-many relations using parameters

### DISCUSSION THREAD

**Discussion by ravargash** [hasura/graphql-engine#9038] [Question] [answered]
Hello everyone!

I've some experience with GraphQL and I'm using typescript, however, I cannot grasp how to do this.

I made a db similar to this:

![Screen Shot 2022-09-29 at 22 25 09](https://user-images.githubusercontent.com/27013133/193170077-0fa6048d-fab7-4a6f-88b7-edeee40fad9c.png)

Both **Services** and **Brands** tables are already filled with items

And I want to create a single mutation that fills every field from **Shops** and include and array of services and brands, filling both **ShopServices** and **ShopBrands** tables. Also, the owner_id is a preset using _x-hasura-user-id_ 

I'm able to generate a mutation doing something like this: 

<img width="299" alt="Screen Shot 2022-09-29 at 22 46 24" src="https://user-images.githubusercontent.com/27013133/193172415-10ead58e-0c45-4aed-9ce5-9c2be678791c.png">

Then, I want to use arguments to be able to use it with a simple fetch function, adding both query and variables into a single "stringified", however, when doing so, I got this

<img width="658" alt="Screen Shot 2022-09-29 at 22 50 42" src="https://user-images.githubusercontent.com/27013133/193172965-d4c95011-f178-4146-8720-229e5d0e0597.png">

I don't know how to tackle both $shop_services and $shop_brands variables nor how should I handle the variables. What should I do?

Please, any help or pointers will be very very appreciated. 

Regards!

**Comment by adas98012** [comment, marked as answer]
We would suggest you to use "Input Types" for complex object types (e.g. shop_services and shop_brands in your case), which can make your mutations simpler to manage. You can refer to the official GraphQL document on the same:

https://graphql.org/learn/schema/#input-types

In your case, you are looking at passing an array of these Input Types (shop_services and shop_brands) as part of the mutation.

Please checkout some illustrations on our Hasura docs page too:
https://hasura.io/docs/latest/mutations/postgres/insert/

**Reply by ravargash** [reply]
Hello!

Thank you. While building the query, I found that Hasura automatically uses the 'mekanika_db_shop_brands_arr_rel_insert_input' and 'mekanika_db_shop_services_arr_rel_insert_input' input types. I left them as they were and then, when filling the variables json, added a new json with the respective shop_brands and shop_services properties. 

This interface I made worked right away:

<img width="266" alt="Screen Shot 2022-10-03 at 20 14 36" src="https://user-images.githubusercontent.com/27013133/193702322-dd8220f5-9863-4bcc-a770-1a67896d1159.png">

Hope it helps

**Reply by adas98012** [reply]
Good to know that you were able to make it work using interface pattern. Please let us know any other issues you encounter in your usage of Hasura GraphQL Engine.
